pkg/storage/azure.go

```diff
@@ -0,0 +1,446 @@
+	}
+
+	prefix := providerStoragePrefix(s.prefix, internalProviderType, "", namespace, name)
+	key := filepath.Join(prefix, filename)
+	return s.upload(ctx, key, file, false)
+}
+
+func (s *AzureStorage) signingKeys(ctx context.Context, pt providerType, hostname, namespace string) (*core.SigningKeys, error) {
+	if namespace == "" {
+		return nil, fmt.Errorf("namespace argument is empty")
+	}
+	key := signingKeysPath(s.prefix, pt, hostname, namespace)
+	exists, err := s.objectExists(ctx, key)
+	if err != nil {
+		return nil, err
+	} else if !exists {
+		return nil, core.ErrObjectNotFound
+	}
+
+	signingKeysRaw, err := s.download(ctx, key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to download signing_keys.json for namespace %s: %w", namespace, err)
+	}
+
+	return unmarshalSigningKeys(signingKeysRaw)
+}
+
+// SigningKeys downloads the JSON placed in the namespace in Azure Blob Storage and unmarshals it into a core.SigningKeys
+func (s *AzureStorage) SigningKeys(ctx context.Context, namespace string) (*core.SigningKeys, error) {
+	return s.signingKeys(ctx, internalProviderType, "", namespace)
+}
+
+func (s *AzureStorage) MirroredSigningKeys(ctx context.Context, hostname, namespace string) (*core.SigningKeys, error) {
+	return s.signingKeys(ctx, mirrorProviderType, hostname, namespace)
+}
+
+func (s *AzureStorage) uploadSigningKeys(ctx context.Context, pt providerType, hostname, namespace string, signingKeys *core.SigningKeys) error {
+	b, err := json.Marshal(signingKeys)
+	if err != nil {
+		return err
+	}
+	key := signingKeysPath(s.prefix, pt, hostname, namespace)
+	return s.upload(ctx, key, bytes.NewReader(b), true)
+}
+
+func (s *AzureStorage) UploadMirroredSigningKeys(ctx context.Context, hostname, namespace string, signingKeys *core.SigningKeys) error {
+	return s.uploadSigningKeys(ctx, mirrorProviderType, hostname, namespace, signingKeys)
+}
+
+func (s *AzureStorage) MirroredSha256Sum(ctx context.Context, provider *core.Provider) (*core.Sha256Sums, error) {
+	prefix := providerStoragePrefix(s.prefix, mirrorProviderType, provider.Hostname, provider.Namespace, provider.Name)
+	key := filepath.Join(prefix, provider.ShasumFileName())
+	shaSumBytes, err := s.download(ctx, key)
+	if err != nil {
+		return nil, errors.New("failed to download SHA256SUMS")
+	}
+
+	return core.NewSha256Sums(provider.ShasumFileName(), bytes.NewReader(shaSumBytes))
+}
+
+func (s *AzureStorage) UploadMirroredFile(ctx context.Context, provider *core.Provider, fileName string, reader io.Reader) error {
+	prefix := providerStoragePrefix(s.prefix, mirrorProviderType, provider.Hostname, provider.Namespace, provider.Name)
+	key := filepath.Join(prefix, fileName)
+	return s.upload(ctx, key, reader, true)
+}
+
+func (s *AzureStorage) presignedURL(ctx context.Context, key string) (string, error) {
+	info := service.KeyInfo{
+		Start:  to.Ptr(time.Now().UTC().Format(sas.TimeFormat)),
+		Expiry: to.Ptr(time.Now().UTC().Add(4 * time.Hour).Format(sas.TimeFormat)),
```

Comment:
Can you explain why this is set to expire in 4h?

Reply:
It's just a random value I choose. We can probably set the time to an even lower value like a couple minutes.

Reply:
Okay, I need to do some more research to understand the implications :)